### PR TITLE
feat: avoid catalog access on each query

### PR DESCRIPTION
**Pull request by alamb**
~Draft as I am breaking this PR up into smaller pieces for easier review. Will mark as ready for review when we need to turn on the code~

~Waiting on https://github.com/influxdata/influxdb_iox/pull/4701~

# Rationale
Closes: https://github.com/influxdata/influxdb_iox/issues/4634

See design doc for more details: https://docs.google.com/document/d/1gszVCQkvY8A40dsNRWJThxdPGV2jsh3u1DolsGpj9XU/edit#heading=h.p96gfd4b8tda

Basically, the catalog (unsurprisingly) is a bottleneck during query time so we shouldn't be querying it over and over again. 

# Changes
This PR adds caching for parquet files and tombstones in the querier along with the appropriate invalidation logic based on the response from the ingesters. 

# PR Chain
To avoid one massive PR I am breaking this one up into several pieces. Here is a list:

- [x] https://github.com/influxdata/influxdb_iox/pull/4644
- [x] https://github.com/influxdata/influxdb_iox/pull/4652
- [x] https://github.com/influxdata/influxdb_iox/pull/4661
- [x] https://github.com/influxdata/influxdb_iox/pull/4663
- [x] https://github.com/influxdata/influxdb_iox/pull/4685
- [x] https://github.com/influxdata/influxdb_iox/pull/4701

# Follow on work related to deletes
- [ ] https://github.com/influxdata/influxdb_iox/issues/4694
- [ ] https://github.com/influxdata/influxdb_iox/issues/4705

# Manual Testing

Here is an end to end manual test showing the behavior we want

I added some logging to the catalog locally:

```diff
diff --git a/iox_catalog/src/postgres.rs b/iox_catalog/src/postgres.rs
index fd2e88519..46679499e 100644
--- a/iox_catalog/src/postgres.rs
+++ b/iox_catalog/src/postgres.rs
@@ -1590,6 +1590,8 @@ WHERE table_id = $1 AND to_delete IS NULL;
         &mut self,
         table_id: TableId,
     ) -> Result<Vec<ParquetFileWithMetadata>> {
+        warn!(%table_id, "Asking postgres for catalog files");
+
         sqlx::query_as::<_, ParquetFileWithMetadata>(
             r#"
 SELECT *
```

Then I ran IOx like this
```shell
LOG_FILTER=warn nice cargo run -- run all-in-one --catalog-dsn=postgresql://localhost:5432/alamb
```

And ran 5 queries like this in another shell
```
influxdb_iox query 26f7e5a4b7be365b_917b97a92e883afc 'select count(*) from cpu'
```

On master there are 5 queries to the catalog
```shell
(arrow_dev) alamb@MacBook-Pro-3:~/Software/influxdb_iox2$ LOG_FILTER=warn nice cargo run -- run all-in-one --catalog-dsn=postgresql://localhost:5432/alamb
    Finished dev [unoptimized + debuginfo] target(s) in 0.39s
     Running `target/debug/influxdb_iox run all-in-one '--catalog-dsn=postgresql://localhost:5432/alamb'`
2022-05-20T15:56:59.712142Z  WARN iox_catalog::postgres: Asking postgres for catalog files table_id=4
2022-05-20T15:57:00.418733Z  WARN iox_catalog::postgres: Asking postgres for catalog files table_id=4
2022-05-20T15:57:02.730330Z  WARN iox_catalog::postgres: Asking postgres for catalog files table_id=4
2022-05-20T15:57:03.223153Z  WARN iox_catalog::postgres: Asking postgres for catalog files table_id=4
2022-05-20T15:57:04.718184Z  WARN iox_catalog::postgres: Asking postgres for catalog files table_id=4
```

On this branch all 5 queries resulted in a single catalog access

```shell
(arrow_dev) alamb@MacBook-Pro-3:~/Software/influxdb_iox$ LOG_FILTER=warn nice cargo run -- run all-in-one --catalog-dsn=postgresql://localhost:5432/alamb
    Finished dev [unoptimized + debuginfo] target(s) in 0.37s
     Running `target/debug/influxdb_iox run all-in-one '--catalog-dsn=postgresql://localhost:5432/alamb'`
2022-05-20T15:58:43.660204Z  WARN iox_catalog::postgres: Asking postgres for catalog files table_id=4
```

### CODE REVIEW

File: querier/src/cache/mod.rs [hunk 1]
```diff
@@ -37,7 +38,13 @@ pub struct CatalogCache {
     namespace_cache: NamespaceCache,
 
     /// Processed tombstone cache.
-    processed_tombstones: ProcessedTombstonesCache,
+    processed_tombstones_cache: ProcessedTombstonesCache,
```
Comment (alamb) [inline querier/src/cache/mod.rs:41]: Here are the new caches added to `CatalogCache`

File: querier/src/cache/mod.rs [hunk 2]
```diff
@@ -125,6 +148,16 @@ impl CatalogCache {
...
-        &self.processed_tombstones
+        &self.processed_tombstones_cache
+    }
+
+    /// Parquet file cache.
+    pub(crate) fn parquet_file(&self) -> &ParquetFileCache {
+        &self.parquet_file_cache
+    }
+
+    /// Parquet file cache.
```
Comment (alamb) [inline querier/src/cache/mod.rs:159]:
```suggestion
    /// tombstone cache.
```

File: querier/src/table/state_reconciler.rs [hunk 3]
```diff
@@ -64,8 +65,8 @@ impl Reconciler {
     pub(crate) async fn reconcile(
         &self,
         ingester_partitions: Vec<IngesterPartition>,
-        tombstones: Vec<Tombstone>,
-        parquet_files: Vec<ParquetFileWithMetadata>,
+        tombstones: Vec<Arc<Tombstone>>,
```
Comment (alamb) [inline querier/src/table/state_reconciler.rs:68]: The changes in this file are to add some debugging as well as change the types to use the Cached variants

File: querier/src/table/test_util.rs [hunk 4]
```diff
@@ -38,11 +41,88 @@ pub async fn querier_table(catalog: &Arc<TestCatalog>, table: &Arc<TestTable>) -
...
 }
+
+/// Convert the line protocol in `lp `to a RecordBatch
+pub(crate) fn lp_to_record_batch(lp: &str) -> RecordBatch {
+    lp_to_mutable_batch(lp).1.to_arrow(Selection::All).unwrap()
+}
+
+/// Helper for creating IngesterPartitions
+#[derive(Debug, Clone)]
+pub(crate) struct IngesterPartitionBuilder {
```
Comment (alamb) [inline querier/src/table/test_util.rs:59]: I plan to refactor the other tests in table/mod.rs to use this builder in a follow on PR, but I didn't want to do so in this PR as it is large enough already

File: querier/src/table/mod.rs [hunk 5]
```diff
@@ -128,49 +123,49 @@ impl QuerierTable {
     pub async fn chunks(&self, predicate: &Predicate) -> Result<Vec<Arc<dyn QueryChunk>>> {
         debug!(?predicate, namespace=%self.namespace_name, table_name=%self.table_name(), "Fetching all chunks");
 
-        // ask ingesters for data
-        let ingester_partitions = self.ingester_partitions(predicate).await?;
+        let catalog_cache = self.chunk_adapter.catalog_cache();
+
+        // ask ingesters for data, optimistically fetching catalog contents at the same time
```
Comment (alamb) [inline querier/src/table/mod.rs:128]: Core change 1. Fetch the data from ingesters and caches in "parallel" (don't wait for ingester to respond before asking catalog)

File: querier/src/table/mod.rs [hunk 6]
```diff
@@ -128,49 +123,49 @@ impl QuerierTable {
...
-                    let tombstones = txn.tombstones().list_by_table(self.id).await?;
-
-                    txn.commit().await?;
-
-                    Ok((parquet_files, tombstones))
-                },
-            )
-            .await
-            .expect("retry forever");
+        let (parquet_files, tombstones) = join!(
```
Comment (alamb) [inline querier/src/table/mod.rs:150]: Core change 2. Refresh cache if necessary

File: querier/src/table/mod.rs [hunk 7]
```diff
@@ -669,13 +740,7 @@ mod tests {
...
                     partition1.partition.id,
                     sequencer.sequencer.id,
-                    Arc::new(
-                        SchemaBuilder::new()
-                            .influx_field("foo", InfluxFieldType::Integer)
-                            .timestamp()
-                            .build()
-                            .unwrap(),
-                    ),
+                    Arc::clone(&schema),
```
Comment (alamb) [inline querier/src/table/mod.rs:743]: drive by cleanup unrelated to the rest of this PR

File: querier/src/table/mod.rs [hunk 8]
```diff
@@ -734,7 +787,152 @@ mod tests {
         assert_matches!(err, Error::IngestersOverlap { .. });
     }
 
-    fn lp_to_record_batch(lp: &str) -> RecordBatch {
-        lp_to_mutable_batch(lp).1.to_arrow(Selection::All).unwrap()
+    #[tokio::test]
+    async fn test_parquet_cache_refresh() {
```
Comment (alamb) [inline querier/src/table/mod.rs:791]: new unit tests

File: influxdb_iox/tests/end_to_end_cases/querier.rs [hunk 9]
```diff
@@ -124,6 +124,89 @@ async fn basic_no_ingester_connection() {
     .await
 }
 
+#[tokio::test]
+async fn query_after_persist_sees_new_files() {
```
[the remark this answers is not in the document]
Comment (alamb) [reply influxdb_iox/tests/end_to_end_cases/querier.rs:128]: End to end test for deletes is tracked in https://github.com/influxdata/influxdb_iox/issues/4694

**Comment by alamb** [comment]
>  BTW: you could (but don't have to) implement an end2end test for the caching behavior by looking at the catalog access metrics (can be retrieved via GET /metrics. I think we don't have a parser for the prometheus metrics yet in our code-base, so either you add one or do manual string-based assertions. Another way to test it would be to add a debug/trace log line and start a non-shared test cluster w/ specific log filters (set via env vars).


There is some coverage of "cache doesn't ask the catalog" in https://github.com/influxdata/influxdb_iox/blob/main/querier/src/cache/parquet_file.rs but I agree an end to end test would be a good idea

🤔  I think OG had some way to look at metrics in the end to end test. I'll see what I can do.

**Review by crepererum — COMMENTED** [review]
I'm not entirely happy w/ the `..._outdated` flags because they replicate logic that is part of the cache layer itself and it seems like premature optimization.

I've left a few testing notes, these are optional however.

File: querier/src/table/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (alamb) [reply querier/src/table/mod.rs:771]: I was thinking that this check (that a subsequent request doesn't pick up the new file that is in the catalog) shows that the data was cached

```rust

        // With the same ingester response, still expect 2 chunks: one
        // for ingester, and one from parquet file
        let chunks = chunks_with_ingester_partition(&querier_table, &ingester_partition).await;
        assert_eq!(chunks.len(), 2);
```

**Comment by alamb** [comment]
> BTW: you could (but don't have to) implement an end2end test for the caching behavior by looking at the catalog access metrics (can be retrieved via GET /metrics. I think we don't have a parser for the prometheus metrics yet in our code-base, so either you add one or do manual string-based assertions. Another way to test it would be to add a debug/trace log line and start a non-shared test cluster w/ specific log filters (set via env vars).

I started doing it here: https://github.com/influxdata/influxdb_iox/pull/4702 but I couldn't figure out what metric to check at this level 🤔

File: influxdb_iox/tests/end_to_end_cases/querier.rs [hunk 10]
```diff
@@ -124,6 +124,89 @@ async fn basic_no_ingester_connection() {
...
+    // https://github.com/influxdata/influxdb_iox/issues/4634 added
+    // caching of tombstones and parquet files in the querier. This
+    // test ensures that a query issued after new parquet files is
+    // issued, they are correctly picked up
```
Comment (carols10cents) [inline influxdb_iox/tests/end_to_end_cases/querier.rs:132]: I'm not sure what this comment is saying, maybe it's supposed to say:

```suggestion
    // test ensures that a query issued after new parquet files are
    // persisted correctly picks up the new parquet files
```

?

File: influxdb_iox/tests/end_to_end_cases/querier.rs [hunk 11]
```diff
@@ -124,6 +124,89 @@ async fn basic_no_ingester_connection() {
...
+        .await
+        .with_ingester(ingester_config)
+        .await
+        .with_querier(querier_config)
+        .await;
+
+    // We need a trigger for persistence that is not time so the test
+    // is as stable as possible. We use a long string to cross the
+    // persistence memory threshold.
+    let super_long_string: String = "x".repeat(10_000);
```
Comment (carols10cents) [inline influxdb_iox/tests/end_to_end_cases/querier.rs:155]: It'd be nice if in a future PR this code was deduplicated with the similar-but-slightly-different-for-some-reason code [elsewhere in this file](https://github.com/influxdata/influxdb_iox/blob/f645ec8a428bb300504954a53e8e4910441f16d4/influxdb_iox/tests/end_to_end_cases/querier.rs#L297-L302), maybe even extracting this into a `Step` like `Step::WriteSuperLongLineProtocolToForcePersistence` or similar?

File: querier/src/cache/mod.rs (hunk 2 above)
Comment (carols10cents) [inline querier/src/cache/mod.rs:159]: Copypasta comment :)

```suggestion
    /// Tombstone cache.
```

File: querier/src/table/mod.rs [hunk 12]
```diff
@@ -128,49 +123,38 @@ impl QuerierTable {
...
-                    let tombstones = txn.tombstones().list_by_table(self.id).await?;
-
-                    txn.commit().await?;
-
-                    Ok((parquet_files, tombstones))
-                },
-            )
-            .await
-            .expect("retry forever");
+        // Now fetch the actual contents of hte catalog we care need
```
Comment (carols10cents) [inline querier/src/table/mod.rs:150]: Comment nit

```suggestion
        // Now fetch the actual contents of the catalog we need
```

File: querier/src/table/mod.rs [hunk 13]
```diff
@@ -734,7 +771,152 @@ mod tests {
...
+        let querier_table = querier_table(&catalog, &table).await;
+        let builder = IngesterPartitionBuilder::new(
+            ns,
+            Arc::clone(&table),
+            schema,
+            Arc::clone(&sequencer),
+            Arc::clone(&partition),
+        );
+
+        // arquet file with max sequence number 1
```
Comment (carols10cents) [inline querier/src/table/mod.rs:841]: Comment nit

```suggestion
        // parquet file with max sequence number 1
```

**Review by carols10cents — APPROVED** [review]
Some comment nits and future refactoring ideas, nothing blocking!

File: querier/src/cache/mod.rs (hunk 2 above)
Comment (alamb) [reply querier/src/cache/mod.rs:159]: 🤦

File: influxdb_iox/tests/end_to_end_cases/querier.rs (hunk 11 above)
Comment (alamb) [reply influxdb_iox/tests/end_to_end_cases/querier.rs:155]: @crepererum originally wrote this test and then I copied it here in parallel. I will consolidate in a follow on PR

Comment (alamb) [reply influxdb_iox/tests/end_to_end_cases/querier.rs:155]: In https://github.com/influxdata/influxdb_iox/pull/4708

**Pull request merged**